File: dart/lib/src/twitter/delete_twitter_registry.dart
import '../constants/addresses.dart';
import '../instructions/delete_name_registry_instruction.dart';
import '../instructions/instruction_types.dart';
import '../utils/get_domain_key_sync.dart';

/// Delete Twitter registry instruction
///
/// This function mirrors js/src/twitter/deleteTwitterRegistry.ts
///
/// Delete the verified registry for a given Twitter handle.
/// Must be signed by the verified pubkey.
///
/// [twitterHandle] - The Twitter handle to delete the registry for
/// [verifiedPubkey] - The verified public key that owns the registry
///
/// Returns a list of transaction instructions
Future<List<TransactionInstruction>> deleteTwitterRegistry(
  String twitterHandle,
  String verifiedPubkey,
) async {
  // Hash the Twitter handle
  final hashedTwitterHandle = getHashedNameSync(twitterHandle);

  // Get the Twitter handle registry key
  final twitterHandleRegistryKey = getNameAccountKeySync(
    hashedTwitterHandle,
    nameParent: twitterRootParentRegistryAddress,
  );

  // Hash the verified pubkey for reverse registry
  final hashedVerifiedPubkey = getHashedNameSync(verifiedPubkey);

  // Get the reverse registry key
  final reverseRegistryKey = getNameAccountKeySync(
    hashedVerifiedPubkey,
    nameClass: twitterVerificationAuthority,
    nameParent: twitterRootParentRegistryAddress,
  );

  final instructions = <TransactionInstruction>[];

  // Delete the user facing registry
  final deleteUserFacingInstr = DeleteNameRegistryInstruction();
  instructions.add(deleteUserFacingInstr.getInstruction(
    programAddress: nameProgramAddress,
    domainAddress: await twitterHandleRegistryKey,
    refundTarget: verifiedPubkey,
    owner: verifiedPubkey,
  ));

  // Delete the reverse registry
  final deleteReverseInstr = DeleteNameRegistryInstruction();
  instructions.add(deleteReverseInstr.getInstruction(
    programAddress: nameProgramAddress,
    domainAddress: await reverseRegistryKey,
    refundTarget: verifiedPubkey,
    owner: verifiedPubkey,
  ));

  return instructions;
}

File: js/src/twitter/deleteTwitterRegistry.ts
import { PublicKey, TransactionInstruction } from "@solana/web3.js";
import {
  NAME_PROGRAM_ID,
  TWITTER_VERIFICATION_AUTHORITY,
  TWITTER_ROOT_PARENT_REGISTRY_KEY,
} from "../constants";
import { deleteInstruction } from "../instructions/deleteInstruction";
import { getHashedNameSync } from "../utils/getHashedNameSync";
import { getNameAccountKeySync } from "../utils/getNameAccountKeySync";

/**
 * Builds instructions to delete a verified Twitter handle registry and reverse registry.
 *
 * The verified public key must sign the resulting instructions.
 *
 * @param twitterHandle Verified Twitter handle whose registries are deleted.
 * @param verifiedPubkey Signer that owns both registries and receives their lamports.
 * @returns Instructions that delete the user-facing and reverse registries.
 *
 * @example
 * ```ts
 * const instructions = await deleteTwitterRegistry("bonfida", owner);
 * ```
 */
export async function deleteTwitterRegistry(
  twitterHandle: string,
  verifiedPubkey: PublicKey,
): Promise<TransactionInstruction[]> {
  const hashedTwitterHandle = getHashedNameSync(twitterHandle);
  const twitterHandleRegistryKey = getNameAccountKeySync(
    hashedTwitterHandle,
    undefined,
    TWITTER_ROOT_PARENT_REGISTRY_KEY,
  );

  const hashedVerifiedPubkey = getHashedNameSync(verifiedPubkey.toString());
  const reverseRegistryKey = getNameAccountKeySync(
    hashedVerifiedPubkey,
    TWITTER_VERIFICATION_AUTHORITY,
    TWITTER_ROOT_PARENT_REGISTRY_KEY,
  );

  const instructions = [
    // Delete the user facing registry
    deleteInstruction(
      NAME_PROGRAM_ID,
      twitterHandleRegistryKey,
      verifiedPubkey,
      verifiedPubkey,
    ),
    // Delete the reverse registry
    deleteInstruction(
      NAME_PROGRAM_ID,
      reverseRegistryKey,
      verifiedPubkey,
      verifiedPubkey,
    ),
  ];

  return instructions;
}
